### PR TITLE
Allow read operations to be executed without waiting for full range to be written in cache

### PR DESCRIPTION
**Pull request by tlrx**
This pull request changes `CacheFile` and `CachedBlobContainerIndexInput` so that the read operations made by these classes are now progressively executed and do not wait for full range to be written in cache. It relies on the change introduced in #58477 and it is the last change extracted from #58164.

Relates #58164

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 1]
```diff
@@ -317,6 +318,14 @@ public CacheFile getCacheFile(CacheKey cacheKey, long fileLength) throws Excepti
         return cacheService.get(cacheKey, fileLength, cacheDir);
     }
 
+    public Executor executor() {
+        return threadPool.executor(SearchableSnapshotsConstants.SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME);
+    }
+
+    public Executor directExecutor() {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:325]: Could also be named `prewarmExecutor()`

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 2]
```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
...
+    CompletableFuture<Integer> fetchAsync(
+        final Tuple<Long, Long> rangeToWrite,
+        final Tuple<Long, Long> rangeToRead,
+        final CacheReader reader,
+        final CacheWriter writer,
+        final Executor executor
     ) {
         final CompletableFuture<Integer> future = new CompletableFuture<>();
         try {
-            if (start < 0 || start > tracker.getLength() || start > end || end > tracker.getLength()) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:270]: This is already verified by the SparseFileTracker itself

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 3]
```diff
@@ -141,13 +143,22 @@ protected void readInternal(ByteBuffer b) throws IOException {
...
-                        range.v2(),
-                        (start, end) -> readCacheFile(cacheFile.getChannel(), end, pos, b, len),
-                        (start, end) -> writeCacheFile(cacheFile.getChannel(), start, end)
-                    ).get();
+                    final Tuple<Long, Long> rangeToWrite = computeRange(pos);
+                    final Tuple<Long, Long> rangeToRead = Tuple.tuple(pos, Math.min(pos + len, rangeToWrite.v2()));
+
+                    bytesRead = cacheFile.fetchAsync(rangeToWrite, rangeToRead, (channel) -> {
+                        final int read;
+                        if ((rangeToRead.v2() - rangeToRead.v1()) < b.remaining()) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:151]: This was previously embedded in the `readCacheFile()` method but feels wrong to me; the `readCacheFile()` should expect a `ByteBuffer` with an appropriate limit.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java [hunk 4]
```diff
@@ -224,31 +235,33 @@ public void prefetchPart(final int part) throws IOException {
...
-                                fileInfo.physicalName(),
-                                writtenBytes
-                            );
-                            totalBytesWritten.addAndGet(writtenBytes);
-                        });
+                        final long readEnd = readStart + bytesRead;
+
+                        cacheFile.fetchAsync(
+                            Tuple.tuple(readStart, readEnd),
+                            Tuple.tuple(readStart, readStart),
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:244]: Maybe it deserves some comment: we don't care about reading the cached data for prewarming, it just need to cache the range. This is why it got passed an empty range to read. This way if the range is already (or about to be) written by a concurrent search it returns immediately and process the next small range. If the range is not available then the cache data will be written by this prewarming task.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java (hunk 3 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java:151]: ++ there was a recent change to the Lucene APIs replacing `byte[], int, int` with `ByteBuffer` but this was not pushed all the way through this code.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 5]
```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
...
             final List<SparseFileTracker.Gap> gaps = tracker.waitForRange(
-                Tuple.tuple(start, end),
-                Tuple.tuple(start, end), // TODO use progressive sub range to trigger read operations sooner
-                ActionListener.wrap(
-                    rangeReady -> future.complete(onRangeAvailable.apply(start, end)),
-                    rangeFailure -> future.completeExceptionally(rangeFailure)
-                )
+                rangeToWrite,
+                rangeToRead,
+                ActionListener.wrap(success -> future.complete(reader.read(channel)), future::completeExceptionally)
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:288]: Can we assert that `reader.read(channel)` returns `rangeToRead.end() - rangeToRead.start()` here?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 6]
```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
...
-        CheckedBiFunction<Long, Long, Integer, IOException> onRangeAvailable,
-        CheckedBiConsumer<Long, Long, IOException> onRangeMissing
+    @FunctionalInterface
+    interface CacheReader {
+        int read(FileChannel channel) throws IOException;
+    }
+
+    @FunctionalInterface
+    interface CacheWriter {
+        void write(FileChannel channel, long from, long to, Consumer<Long> progressUpdater) throws IOException;
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:272]: What, no `CheckedQuadConsumer<...>`? 😁 

I'm torn about having these be separate interfaces vs combining them together perhaps along with some other arguments to `fetchAsync` too. I'm finding it especially strange that `CacheReader#read` only takes a `FileChannel`; this indicates that `rangeToRead` is now available, but that's very implicit now.

Maybe different names would help. How about something like `RangeMissingHandler#fillCacheRange` and `RangeAvailableHandler#onRangeAvailable`?

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I suggested some small cosmetic changes but this seems structurally sound.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:325]: > Can we name the other executor() method something more specific too?

I'm terrible at naming things (sorry, kids)... what do you think of ` searchExecutor()` ? `cacheWritingExecutor()` ?

Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:325]: Tricky one TBH. How about `cacheFetchAsyncExecutor()`?

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:325]: +1

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java [hunk 7]
```diff
@@ -259,38 +262,68 @@ private void ensureOpen() {
...
+                                    }
+                                    writer.write(channel, gap.start(), gap.end(), gap::onProgress);
+                                    gap.onCompletion();
+                                } finally {
+                                    readLock.unlock();
+                                }
+                            } catch (Exception e) {
+                                gap.onFailure(e);
+                            } finally {
+                                iterator.remove();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:317]: This is a left over - I think we can just iterate over the list of gaps.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java (hunk 5 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:288]: Good suggestion, thanks

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java (hunk 6 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java:272]: > What, no CheckedQuadConsumer<...>? grin

Don't tempt me :)



> Maybe different names would help. How about something like `RangeMissingHandler#fillCacheRange` and `RangeAvailableHandler#onRangeAvailable`?

I went this route and renamed to RangeMissingHandler/RangeAvailableHandler

**Comment by tlrx** [comment]
Thanks @DaveCTurner - I updated the code with your comments. I also added another assertion in the positionalWrite() method.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**